### PR TITLE
merge.GADSdat: Add automatic recoding of NAs

### PR DESCRIPTION
**Pull request by beckerbenj**
This feature addresses #1. If two `GADsdat` objects are merged, it is desirable to automatically assign a `missingValue` and a `missingValLabel` to the `NAs` created through merging.

I am not super happy with my implementation. The workaround using `intermediate_code` seems vulnerable and bad practice to me. However, I was unable to come up with a working and somewhat efficient alternative. Would be very happy to hear your thoughts on this.

Feature requested for BT24 data cleaning (@liebelta, @OtteKatrin, @burbliesj)

### CODE REVIEW

File: R/merge.R [hunk 1]
```diff
@@ -2,37 +2,78 @@
...
+    stop(by, " is not a character vector.")
+  }
+  check_vars_in_GADSdat(x, vars = by, argName = "by", GADSdatName = "x")
+  check_vars_in_GADSdat(y, vars = by, argName = "by", GADSdatName = "y")
+
+  # drop duplicate variables from y
   y_vars <- c(by, names(y$dat)[!names(y$dat) %in% names(x$dat)])
-  if(!length(y_vars) > length(by)) stop("y does not contain unique variables.")
+  if(!length(y_vars) > length(by)) {
+    stop("y does not contain unique variables.")
```
Comment (franikowsp) [inline R/merge.R:41]: Nit: I was a little bit confused by this error message as I was thinking of y names like `c("ID", "V1", "V1", "V1")`. Maybe, this would be a bit more helpful: `"y does not contain variables distinct from x."`? 🤔

File: tests/testthat/test_merge.R [hunk 2]
```diff
@@ -1,32 +1,45 @@
...
+  df1b$dat[1, "V1"] <- NA
+  df3b <- df3
+  df3b$dat[2:3, "V2"] <- NA
+  out2 <- merge(df1b, df3b, by = "ID1", missingValue = -99, missingValLabel = "missing")
+  expect_equal(out2$dat[, "V1"], c(NA, 5, -99, -99))
+  expect_equal(out2$dat[, "V2"], c(8, -99, NA, NA))
+  expect_equal(nrow(out2$labels), 3)
+  expect_equal(out2$labels[2:3, "value"], c(-99, -99))
+  expect_equal(out2$labels[2:3, "valLabel"], c("missing", "missing"))
+  expect_equal(out2$labels[2:3, "missings"], c("miss", "miss"))
```
Comment (franikowsp) [inline tests/testthat/test_merge.R:39]: Nit: Maybe, one test battery could be added where `missingValLabel `is `NULL` (I first thought, the `missingValLabel` was required when `missingValue` was set):

```r
  # with existing NAs without valLabel
  df1c <- df1
  df1c$dat[1, "V1"] <- NA
  df3c <- df3
  df3c$dat[2:3, "V2"] <- NA
  out3 <- merge(df1c, df3c, by = "ID1", missingValue = -99)
  expect_equal(out3$dat[, "V1"], c(NA, 5, -99, -99))
  expect_equal(out3$dat[, "V2"], c(8, -99, NA, NA))
  expect_equal(nrow(out3$labels), 3)
  expect_equal(out3$labels[2:3, "value"], c(-99, -99))
  expect_equal(out3$labels[2:3, "valLabel"], c(NA_character_, NA_character_))
  expect_equal(out3$labels[2:3, "missings"], c("miss", "miss"))
```

File: R/merge.R [hunk 3]
```diff
@@ -2,37 +2,78 @@
...
 
-  newDat <- merge(x$dat, y$dat[, y_vars], by = by, all = all, all.x = all.x, all.y = all.y)
-  newLabels <- rbind(x$labels, y$labels[y$labels$varName %in% y_vars[!y_vars %in% by], ])
+  # replace NAs in original data sets to be able to recode them after merging
+  if(!is.null(missingValue)) {
+    check_numericArgument(missingValue, argName = "missingValue")
+
+    intermediate_code <- -99.9989
+    if(any(!is.na(x$dat) & x$dat == intermediate_code) || any(!is.na(y$dat) & y$dat == intermediate_code)) {
+      stop("Intermediate code is already in the data.")
```
Comment (franikowsp) [inline R/merge.R:50]: Hmmm... What would the user do in this case? 🤔 I guess it is highly improbable that the user would have such a code in their dataset. But maybe, there could be a solution offered, e.g., `"... Please replace entries of -99.9989 from your dataset"`? (Does this error message also require a test to be sure it fires somewhen?)

```r
 # with existing NAs without valLabel
  df1c <- df1
  df1c$dat[1, "V1"] <-  -99.9989
  df3c <- df3
  df3c$dat[2:3, "V2"] <- NA
  out3 <- merge(df1c, df3c, by = "ID1", missingValue = -99)
  expect_error(merge(df1c, df3c, by = "ID1", missingValue = -99), "Intermediate code is already in the data\\.")
```

Comment (franikowsp) [reply R/merge.R:50]: Alternatively, for this highly improbable case: Could it maybe be packed in some sort of loop that adds more digits (up to 10)? 🤔 (But this might be over-engineering!)

File: R/merge.R [hunk 4]
```diff
@@ -2,37 +2,78 @@
...
+      newGADS <- changeMissings(newGADS, varName = single_var, value = missingValue, missings = "miss")
+    }
+
+    if(!is.null(missingValLabel) && length(vars_with_missingValue) > 0) {
+      check_characterArgument(missingValLabel, argName = "missingValLabel")
+      for(single_var in vars_with_missingValue) {
+        newGADS <- changeValLabels(newGADS, varName = single_var, value = missingValue, valLabel = missingValLabel)
+      }
+    }
+  }
```
Comment (franikowsp) [inline R/merge.R:77]: Looks very good to me.

**Review by franikowsp — APPROVED** [review]
Sorry for the belated code review. 

LAllVeryGTM. 

**Nit summary:** Maybe, some tests for edge cases could be added (see my proposals) and the error messages could be slightly more informative.

File: R/merge.R (hunk 3 above)
Comment (franikowsp) [reply R/merge.R:50]: Just asked AI and it also came up with a loop solution (from my opinion, it should not let the execution take longer as only a limited range of missings would be plausible):

```r
  existing_values <- c(x$dat, y$dat)
  intermediate_code <- -99.9989  # Default value
  while (any(!is.na(existing_values) & existing_values == intermediate_code)) {
    intermediate_code <- runif(1, -1e6, 1e6)  # Generate a random value in a wide range
  }
```

File: tests/testthat/test_merge.R (hunk 2 above)
Comment (beckerbenj) [reply tests/testthat/test_merge.R:39]: Agree and modified as suggested (but with some simplified code, re-using the data of the previous test case)!

File: R/merge.R (hunk 3 above)
Comment (beckerbenj) [reply R/merge.R:50]: I have implemented your solution as suggested. Thanks for taking the time and coming up with these alternatives!

Only tweaks I made to your suggestions:
- I pre-check for any `NAs` in the data now, to avoid performing any memory or time consuming computations 
- I avoid creating a vector `existing_values`, as for large data sets this could be memory intensive

**Comment by beckerbenj** [comment]
Thanks for the (as always) super helpful review!
